From 82ea2ad236591be9deb92d655a29c53774228b88 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Wed, 29 Mar 2017 13:32:17 -0700
Subject: [PATCH 1488/1566] drm: xilinx: dp: Retry the aux transaction when it
 fails

commit  7e2a1b55be8f772a2ef402354e228c42ef49d736 from
https://github.com/Xilinx/linux-xlnx.git

Retry the aux transaction up to 128 times. Per spec, it needs to
try at least 3 times, but there's some monitor (Viewsonic VX2880ml)
which requires 64 retries. Try up to 128 times to be safe.
Other monitors are more likely to succeed for the first try, so this
patch doesn't hurt other monitors.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_dp.c |   18 +++++++++++++-----
 1 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
index eeed0c5..952d8cd 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
@@ -1229,13 +1229,21 @@ xilinx_drm_dp_aux_transfer(struct drm_dp_aux *aux, struct drm_dp_aux_msg *msg)
 {
 	struct xilinx_drm_dp *dp = container_of(aux, struct xilinx_drm_dp, aux);
 	int ret;
+	unsigned int i;
 
-	ret = xilinx_drm_dp_aux_cmd_submit(dp, msg->request, msg->address,
-					   msg->buffer, msg->size, &msg->reply);
-	if (ret < 0)
-		return ret;
+	for (i = 0; i < 128; i++) {
+		ret = xilinx_drm_dp_aux_cmd_submit(dp, msg->request,
+						   msg->address, msg->buffer,
+						   msg->size, &msg->reply);
+		if (!ret)
+			return msg->size;
 
-	return msg->size;
+		usleep_range(400, 500);
+	}
+
+	dev_dbg(dp->dev, "failed to do aux transfer (%d)\n", ret);
+
+	return ret;
 }
 
 static int xilinx_drm_dp_parse_of(struct xilinx_drm_dp *dp)
-- 
1.7.5.4

